impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java

```diff
@@ -47,140 +59,134 @@ public class KafkaClusterTest {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
+    @BeforeEach
+    public void initializeIndexes() {
+        if (clusterSizesIndex == clusterSizes().count()) {
+            clusterSizesIndex = 0;
         }
     }
 
-    @Test
-    public void kafkaClusterKraftModeWithAuth() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .kraftMode(true)
-                .testInfo(testInfo)
-                .securityProtocol("SASL_PLAINTEXT")
-                .saslMechanism("PLAIN")
-                .user("guest", "guest")
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    @TestTemplate
+    @ExtendWith(KafkaClusterExtension.class)
+    public void kafkaClusterTemplate(@ConstraintsMethodSource("clusterSizesAndModes") InVMKafkaCluster cluster)
+            throws Exception {
+        verifyRecordRoundTrip(((BrokerCluster) brokerClusters[clusterSizesIndex++]).numBrokers(), cluster);
     }
 
-    @Test
-    public void kafkaClusterZookeeperModeWithAuth() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(false)
-                .securityProtocol("SASL_PLAINTEXT")
-                .saslMechanism("PLAIN")
-                .user("guest", "guest")
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    @TestTemplate
+    @ExtendWith(KafkaClusterExtension.class)
+    public void kafkaClusterContainerKraftModeTemplate(@DimensionMethodSource("versions") @DimensionMethodSource("clusterSizes") @KRaftCluster TestcontainersKafkaCluster cluster)
+            throws Exception {
+        verifyRecordRoundTrip(((BrokerCluster) brokerClusters[clusterSizesIndex++]).numBrokers(), cluster);
     }
 
-    @Test
-    public void kafkaClusterKraftModeSASL_SSL() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .keytoolCertificateGenerator(keytoolCertificateGenerator)
-                .kraftMode(true)
-                .securityProtocol("SASL_SSL")
-                .saslMechanism("PLAIN")
-                .user("guest", "guest")
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    @TestTemplate
+    @ExtendWith(KafkaClusterExtension.class)
+    public void kafkaClusterContainerZookeeperModeTemplate(@DimensionMethodSource("versions") @DimensionMethodSource("clusterSizes") @ZooKeeperCluster TestcontainersKafkaCluster cluster)
```

Comment:
```suggestion
    public void kafkaClusterContainerZookeeperModeTemplate(@DimensionMethodSource("versions") @DimensionMethodSource("clusterSizes") @ZooKeeperCluster KafkaCluster cluster)
```

Again not sure why we need the concrete type in the method signature.